docs/dev/libs/state_processor_api.md

```diff
@@ -0,0 +1,459 @@
+ 
+class ReaderFunction extends KeyedStateReaderFunction[Integer, KeyedState] {
+  var state ValueState[Integer];
+ 
+  override def open(Configuration parameters) {
+     state = getRuntimeContext().getState(stateDescriptor);
+  }
+ 
+  override def processKey(
+    key: Int,
+    ctx: Context,
+    out: Collector[Keyedstate]) throws Exception {
+ 
+     val data = KeyedState(key, state.value())
+     out.collect(data);
+  }
+}
+
+val keyedState = savepoint.readKeyedState("my-uid", new ReaderFunction());
+{% endhighlight %}
+</div>
+</div>
+
+{% panel **Note:** When using a `KeyedStateReaderFunction` all state descriptors must be registered eagerly inside of open. Any attempt to call `RuntimeContext#getState`, `RuntimeContext#getListState`, or `RuntimeContext#getMapState` will result in a `RuntimeException`. %}
+
+## Writing New Savepoints
+
+State writers are based around the abstraction of `Savepoint`, where one `Savepoint` may have many operators and the state for any particular operator is created using a `BootstrapTransformation`.
+
+A `BootstrapTransformation` starts with a `DataSet` containing the values that are to be written into state.
+The transformation may be optionally `keyed` depending on whether or not you are writing keyed or operator state.
+Finally a bootstrap function is applied depending to the transformation; Flink supplies `KeyedStateBootstrapFunction` for writing keyed state, `StateBootstrapFunction` for writing non keyed state, and `BroadcastStateBootstrapFunction` for writing broadcast state.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+public  class Account {
+    public int id;
+
+    public double amount;	
+
+    public long timestamp;
+}
+ 
+public class AccountBootstrapper extends KeyedStateBootstrapFunction<Integer, Account> {
+    ValueState<Double> state;
+
+    @Override
+    public void open(Configuration parameters) {
+        ValueStateDescriptor<Double> descriptor = new ValueStateDescriptor<>("total",Types.DOUBLE);
+        state = getRuntimeContext().getState(descriptor);
+    }
+
+    @Override
+    public void processElement(Account value, Context ctx) throws Exception {
+        state.update(value.amount);
+    }
+}
+ 
+ExecutionEnvironment bEnv = ExecutionEnvironment.getExecutionEnvironment();
+
+DataSet<Account> accountDataSet = bEnv.fromCollection(accounts);
+
+BootstrapTransformation<Account> transformation = OperatorTransformation
+    .bootstrapWith(accountDataSet)
+    .keyBy(acc -> acc.id)
+    .transform(new AccountBootstrapper());
+{% endhighlight %}
+</div>
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+case class Account(id: Int, amount: Double, timestamp: Long)
+ 
+class AccountBootstrapper extends KeyedStateBootstrapFunction[Integer, Account] {
+    var state: ValueState[Double]
+
+    override def open(parameters: Configuration): Unit = {
+        val descriptor = new ValueStateDescriptor[Double]("total",Types.DOUBLE)
```

Comment:
I wonder could this be acceptable for scala? Should the `Double` be `JDouble` after we `import java.lang.{Double => JDouble}`

Reply:
Nice catch